src/stream/src/executor/hash_join.rs

```diff
@@ -821,246 +890,401 @@ impl<K: HashKey, S: StateStore, const T: JoinTypePrimitive> HashJoinExecutor<K,
+        {
+            yield chunk;
+        }
+
+        // cache refill
+        if cache_hit || entry_state_count <= entry_state_max_rows {
+            side_match.ht.update_state(key, entry_state);
+        }
+
+        // watermark state cleaning
+        for matched_row in matched_rows_to_clean {
+            side_match.ht.delete_handle_degree(key, matched_row)?;
+        }
+
+        // apply append_only optimization to clean matched_rows which have been persisted
+        if append_only_optimize && let Some(row) = append_only_matched_row {
+            assert_matches!(JOIN_OP, JoinOp::Insert);
+            side_match.ht.delete_handle_degree(key, row)?;
+            return Ok(());
+        }
+
+        // no append_only optimization, update state table(s).
+        match JOIN_OP {
+            JoinOp::Insert => {
+                side_update
+                    .ht
+                    .insert_handle_degree(key, JoinRow::new(row, degree))?;
+            }
+            JoinOp::Delete => {
+                side_update
+                    .ht
+                    .delete_handle_degree(key, JoinRow::new(row, degree))?;
+            }
+        }
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    #[inline]
+    async fn handle_match_row<
+        'a,
+        const SIDE: SideTypePrimitive,
+        const JOIN_OP: JoinOpPrimitive,
+        const MATCHED_ROWS_FROM_CACHE: bool,
+    >(
+        update_row: RowRef<'a>,
+        mut matched_row: JoinRow<OwnedRow>,
+        mut matched_row_cache_ref: Option<&mut StateValueType>,
+        hashjoin_chunk_builder: &'a mut JoinChunkBuilder<T, SIDE>,
+        match_order_key_indices: &[usize],
+        match_degree_table: &mut Option<TableInner<S>>,
+        side_update_start_pos: usize,
+        side_match_start_pos: usize,
+        cond: &Option<NonStrictExpression>,
+        update_row_degree: &mut u64,
+        useful_state_clean_columns: &[(usize, &'a Watermark)],
+        append_only_optimize: bool,
+        append_only_matched_row: &mut Option<JoinRow<OwnedRow>>,
+        matched_rows_to_clean: &mut Vec<JoinRow<OwnedRow>>,
+    ) -> Option<StreamChunk> {
+        let mut need_state_clean = false;
+        let mut chunk_opt = None;
+
+        // TODO(kwannoel): Instead of evaluating this every loop,
+        // we can call this only if there's a non-equi expression.
+        // check join cond
+        let join_condition_satisfied = Self::check_join_condition(
+            update_row,
+            side_update_start_pos,
+            &matched_row.row,
+            side_match_start_pos,
+            cond,
+        )
+        .await;
+
+        if join_condition_satisfied {
+            // update degree
+            *update_row_degree += 1;
+            // send matched row downstream
+
+            // Inserts must happen before degrees are updated,
+            // FIXME(kwannoel): We should let deletes and inserts happen BEFORE degree updates.
```

Comment:
What if we don't let deletes and inserts happen BEFORE degree updates?

Reply:
This is more like a TODO to improve code quality. Currently handling inserts happens before updating degree table, whereas handling deletes happens after.

The reason for that is the `with_match_on_<OP>` functions make some implicit assumptions.
For `with_match_on_insert`, it assumes the matched row does not have its degree updated yet.
For `with_match_on_delete`, it assumes the matched row already has its degree updated.

This is existing logic. If I refactor it, I will have to add more changes, which are IMO out-of-scope. It will also touch ASOF join code.

So I leave a FIXME here. I will fix it in subsequent PRs.

Reply:
Thanks for the explanation!